src/NuGet.Protocol.Core.v3/ServiceIndexResourceV3Provider.cs

```diff
@@ -113,26 +107,31 @@ public ServiceIndexResourceV3Provider()
                     entryValidCutoff > cacheInfo.CachedTime)
                 {
                     var json = await GetIndexJson(source, token);
-                    if (json != null)
+
+                    // Use SemVer instead of NuGetVersion, the service index should always be
+                    // in strict SemVer format                    
+                    JToken versionToken;
+                    if (json.TryGetValue("version", out versionToken) &&
+                        versionToken.Type == JTokenType.String)
                     {
-                        // Use SemVer instead of NuGetVersion, the service index should always be
-                        // in strict SemVer format
                         SemanticVersion version;
-                        JToken versionToken;
-                        if (json.TryGetValue("version", out versionToken) &&
-                            versionToken.Type == JTokenType.String &&
-                            SemanticVersion.TryParse((string)versionToken, out version) &&
+                        if (SemanticVersion.TryParse((string)versionToken, out version) &&
                             version.Major == 3)
                         {
                             index = new ServiceIndexResourceV3(json, utcNow);
                         }
+                        else
+                        {
+                            string errorMessage = string.Format(
+                                CultureInfo.CurrentCulture,
+                                Strings.Protocol_UnsupportedVersion,
+                                (string)versionToken);
+                            throw new NuGetProtocolException(errorMessage);
+                        }
                     }
                     else
                     {
-                        var entry = new ServiceIndexCacheInfo { CachedTime = utcNow };
-                        _cache.AddOrUpdate(url, entry, (key, value) => entry);
-
-                        return new Tuple<bool, INuGetResource>(false, null);
+                        throw new NuGetProtocolException(Strings.Protocol_MissingVersion);
```

Comment:
This throw, plus the previous throw, makes sure that there is no exception only when index.json is legit.